osier/models/dispatch.py

```diff
@@ -125,58 +273,211 @@ def indices(self):
+        for g in self.model.Generators:
+            unit_capacity = (self.capacity_dict[g] * self.time_delta).to_value()
+
+            for t in self.model.Time:
+                unit_generation = self.model.x[g, t]
                 self.model.gen_limit.add(unit_generation <= unit_capacity)
 
+    def _ramping_constraints(self):
+        self.model.ramp_up_limit = pe.ConstraintList()
+        self.model.ramp_down_limit = pe.ConstraintList()
+
+        for r in self.model.RampingTechs:
+            ramp_up = self.model.ramp_up[r]
+            ramp_down = self.model.ramp_down[r]
+            for t in self.model.Time:
+                if t != self.model.Time.first():
+                    t_prev = self.model.Time.prev(t)
+                    previous_gen = self.model.x[r, t_prev]
+                    current_gen = self.model.x[r, t]
+                    delta_power = (current_gen - previous_gen) / \
+                        self.time_delta.to_value()
+                    self.model.ramp_up_limit.add(delta_power <= ramp_up)
+                    self.model.ramp_down_limit.add(delta_power >= -ramp_down)
+
+    def _storage_constraints(self):
+        self.model.discharge_limit = pe.ConstraintList()
+        self.model.charge_limit = pe.ConstraintList()
+        self.model.charge_rate_limit = pe.ConstraintList()
+        self.model.storage_limit = pe.ConstraintList()
+        self.model.set_storage = pe.ConstraintList()
+        for s in self.model.StorageTech:
+            efficiency = self.efficiency_dict[s]
+            storage_cap = self.model.storage_capacity[s]
+            unit_capacity = (self.capacity_dict[s] * self.time_delta).to_value()
+            initial_storage = self.model.initial_storage[s]
+            for t in self.model.Time:
+                self.model.charge_rate_limit.add(
+                    self.model.charge[s, t] <= unit_capacity)
+                if t == self.model.Time.first():
+                    self.model.set_storage.add(self.model.storage_level[s, t]
+                                               == initial_storage)
+                    self.model.discharge_limit.add(
+                        self.model.x[s, t] <= initial_storage)
+                    self.model.charge_limit.add(self.model.charge[s, t]
+                                                <= storage_cap
+                                                - initial_storage)
+                else:
+                    t_prev = self.model.Time.prev(t)
+                    previous_storage = self.model.storage_level[s, t_prev]
+                    current_discharge = self.model.x[s, t]
+                    current_charge = self.model.charge[s, t]
+                    self.model.set_storage.add(
+                        self.model.storage_level[s, t]
+                        == previous_storage
+                        + np.sqrt(efficiency) * current_charge
+                        - np.sqrt(efficiency) * current_discharge
```

Comment:
Because "efficiency" refers to the *round trip* efficiency. Neither charging nor discharging is a lossless activity (I think this is more easily seen if you think about "charging" and "discharging" pumped hydro since there are turbines involved).

Reply:
That makes sense, but that does not make it clear where the square root comes from.

Reply:
Okay here's a concrete example:

If I have 100 MW excess to charge the battery (with a round trip efficiency of 90%) the stored energy without the square root becomes 0.9*100=90. When all of that energy is discharged, the discharged energy is 0.9*90=81. But that doesn't make sense because I told you the round trip efficiency was 0.9 not 0.81! 

The round trip efficiency is the charging efficiency X the discharging efficiency. Since we only specify the round trip efficiency we can only say that charging and discharging are equally efficient. Make sense?